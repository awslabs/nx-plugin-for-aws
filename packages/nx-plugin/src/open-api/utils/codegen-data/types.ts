/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { Plugin } from '@hey-api/openapi-ts';
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type ClientData = Parameters<Plugin.LegacyHandler<any>>[0]['client'];
export type Operation = ClientData['services'][number]['operations'][number];
export type Model = ClientData['models'][number];

export interface CodeGenData extends ClientData {
  className: string;
  info: OpenAPIV3.InfoObject | OpenAPIV3_1.InfoObject;
  allOperations: Operation[];
  operationsByTag: { [tag: string]: Operation[] };
  untaggedOperations: Operation[];
  vendorExtensions: { [key: string]: any };
}

export const flattenModelLink = (link?: Model | Model[]): Model =>
  link === undefined ? undefined : Array.isArray(link) ? link[0] : link;

// Model types which indicate it is composed (ie inherits/mixin's another schema)
export const COMPOSED_SCHEMA_TYPES = new Set(['one-of', 'any-of', 'all-of']);
export const COLLECTION_TYPES = new Set(['array', 'dictionary']);
export const PRIMITIVE_TYPES = new Set([
  'string',
  'integer',
  'number',
  'boolean',
  'null',
  'any',
  'binary',
  'void',
]);
