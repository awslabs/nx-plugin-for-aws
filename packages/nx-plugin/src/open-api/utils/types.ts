/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import type { OpenAPIV3, OpenAPIV3_1 } from 'openapi-types';

export type Spec = OpenAPIV3.Document | OpenAPIV3_1.Document;
