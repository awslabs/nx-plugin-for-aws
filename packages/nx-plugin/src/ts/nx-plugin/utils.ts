/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  addDependenciesToPackageJson,
  joinPathFragments,
  readJson,
  Tree,
  updateJson,
  writeJson,
} from '@nx/devkit';
import { readProjectConfigurationUnqualified } from '../../utils/nx';
import { withVersions } from '../../utils/versions';
import PackageJson from '../../../package.json';

/**
 * Configures a TypeScript project as an Nx Plugin
 */
export const configureTsProjectAsNxPlugin = (
  tree: Tree,
  projectName: string,
) => {
  const project = readProjectConfigurationUnqualified(tree, projectName);

  const tsConfigPath = joinPathFragments(project.root, 'tsconfig.json');
  if (!tree.exists(tsConfigPath)) {
    throw new Error(
      `Selected plugin project ${projectName} is not a TypeScript project`,
    );
  }

  // Nx Plugins must use commonjs as a limitation of nx
  // https://github.com/nrwl/nx/issues/15682
  updateJson(tree, tsConfigPath, (tsConfig) => {
    tsConfig.compilerOptions ??= {};
    tsConfig.compilerOptions.module ??= 'commonjs';
    return tsConfig;
  });

  // Create an empty generators.json if one dosn't exist
  const generatorsJsonPath = joinPathFragments(project.root, 'generators.json');
  if (!tree.exists(generatorsJsonPath)) {
    writeJson(tree, generatorsJsonPath, {
      generators: {},
    });
  }

  // Create a package.json if one doesn't exist, and configure "main" and "generators"
  const pluginPackageJsonPath = joinPathFragments(project.root, 'package.json');
  if (!tree.exists(pluginPackageJsonPath)) {
    writeJson(tree, pluginPackageJsonPath, {
      name: project.name,
    });
  }
  updateJson(tree, pluginPackageJsonPath, (pkg) => {
    pkg.main ??= './src/index.js';
    pkg.generators ??= './generators.json';
    return pkg;
  });

  const rootPackageJson = tree.exists('package.json')
    ? readJson(tree, 'package.json')
    : undefined;
  const isNxPluginForAws = rootPackageJson?.name === '@aws/nx-plugin-source';

  if (!isNxPluginForAws) {
    // Add the required dependencies to the root package.json, and project's package.json
    const deps = {
      ...withVersions(['@nx/devkit']),
      // Include dependency on @aws/nx-plugin
      [PackageJson.name]: `^${PackageJson.version}`,
    };
    addDependenciesToPackageJson(tree, {}, deps);
    addDependenciesToPackageJson(tree, deps, {}, pluginPackageJsonPath);
  }
};
