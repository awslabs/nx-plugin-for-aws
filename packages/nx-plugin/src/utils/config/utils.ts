/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { generateFiles, joinPathFragments, Tree } from '@nx/devkit';
import { AwsNxPluginConfig } from '.';
import * as ts from 'typescript';
import { jsonToAst, replace } from '../ast';
import { factory } from 'typescript';
import { formatFilesInSubtree } from '../format';
import { importTypeScriptModule } from '../js';

export const AWS_NX_PLUGIN_CONFIG_FILE_NAME = 'aws-nx-plugin.config.mts';

/**
 * Ensure that the config file exists
 */
export const ensureAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    // Create an empty config file if it doesn't already exist
    generateFiles(tree, joinPathFragments(__dirname, 'files'), '.', {});
  }
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return await readAwsNxPluginConfig(tree)!;
};

/**
 * Read config from the aws nx plugin configuration file
 */
export const readAwsNxPluginConfig = async (
  tree: Tree,
): Promise<AwsNxPluginConfig | undefined> => {
  if (!tree.exists(AWS_NX_PLUGIN_CONFIG_FILE_NAME)) {
    return undefined;
  }
  const configTs = tree.read(AWS_NX_PLUGIN_CONFIG_FILE_NAME, 'utf-8');
  return await importTypeScriptModule(configTs);
};

/**
 * Update the aws nx plugin config file.
 * Undefined top level keys in the config update will be untouched, otherwise config is replaced
 */
export const updateAwsNxPluginConfig = async (
  tree: Tree,
  configUpdate: Partial<AwsNxPluginConfig>,
): Promise<void> => {
  // Replace the default export
  replace(
    tree,
    AWS_NX_PLUGIN_CONFIG_FILE_NAME,
    'ExportAssignment ObjectLiteralExpression',
    (node) => {
      const existingObj = node as ts.ObjectLiteralExpression;

      const existingProps = new Map<string, ts.PropertyAssignment>();
      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          existingProps.set(prop.name.getText(), prop);
        }
      });

      const properties: ts.PropertyAssignment[] = [];

      for (const [key, value] of Object.entries(configUpdate)) {
        properties.push(
          factory.createPropertyAssignment(
            key,
            jsonToAst(value) as ts.Expression,
          ),
        );
      }

      existingObj.properties.forEach((prop) => {
        if (ts.isPropertyAssignment(prop)) {
          const name = prop.name.getText();
          if (configUpdate[name] === undefined) {
            properties.push(prop);
          }
        }
      });

      return factory.createObjectLiteralExpression(properties, true);
    },
  );

  // Format the config nicely after an update
  await formatFilesInSubtree(tree, AWS_NX_PLUGIN_CONFIG_FILE_NAME);
};
