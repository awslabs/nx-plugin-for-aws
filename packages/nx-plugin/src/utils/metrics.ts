/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, joinPathFragments } from '@nx/devkit';
import { ArrayLiteralExpression, StringLiteral, factory } from 'typescript';
import { replaceIfExists } from './ast';
import { formatFilesInSubtree } from './format';
import { NxGeneratorInfo, getPackageVersion } from './nx';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from './shared-constructs-constants';

// Used to identify @aws/nx-plugin in AWS metrics
export const METRIC_ID = 'uksb-4wk0bqpg5s';

// File in which the MetricsAspect may exist
export const METRICS_ASPECT_FILE_PATH = joinPathFragments(
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  'src',
  'core',
  'app.ts',
);

// Query to find a particular metrics aspect variable
export const metricsAspectVariableQuery = (
  variableName: string,
  valueQuery: string,
) =>
  `ClassDeclaration[name.name="MetricsAspect"] MethodDeclaration[name.name="visit"] VariableDeclaration:has(Identifier[name="${variableName}"]) ${valueQuery}`;

/**
 * Instruments metrics by updating the MetricsAspect in common/constructs/src/core/app.ts if the file exists
 */
export const addGeneratorMetricsIfApplicable = async (
  tree: Tree,
  generatorInfo: NxGeneratorInfo[],
) => {
  if (tree.exists(METRICS_ASPECT_FILE_PATH)) {
    // Update the id
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('id', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(METRIC_ID, true);
      },
    );
    // Update the version
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('version', 'StringLiteral'),
      (): StringLiteral => {
        return factory.createStringLiteral(getPackageVersion(), true);
      },
    );
    // Add each generator as a tag
    replaceIfExists(
      tree,
      METRICS_ASPECT_FILE_PATH,
      metricsAspectVariableQuery('tags', 'ArrayLiteralExpression'),
      (node: ArrayLiteralExpression): ArrayLiteralExpression => {
        const existingMetrics = new Set(
          node.elements.map((element) => (element as StringLiteral)?.text),
        );

        return factory.createArrayLiteralExpression([
          ...node.elements,
          ...generatorInfo
            .filter((info) => !existingMetrics.has(info.metric))
            .map((info) => factory.createStringLiteral(info.metric, true)),
        ]);
      },
    );
    await formatFilesInSubtree(tree, METRICS_ASPECT_FILE_PATH);
  }
};
