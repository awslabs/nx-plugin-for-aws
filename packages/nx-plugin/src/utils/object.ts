/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

export const sortObjectKeys = <T>(object: {
  [key: string]: T;
}): { [key: string]: T } =>
  Object.keys(object)
    .sort()
    .reduce((obj, key) => {
      obj[key] = object[key];
      return obj;
    }, {});
