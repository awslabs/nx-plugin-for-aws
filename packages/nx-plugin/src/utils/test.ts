/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { vi, expect } from 'vitest';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { joinPathFragments, Tree } from '@nx/devkit';
import solutionSetup from '@nx/js/src/utils/typescript/ts-solution-setup';

export const createTreeUsingTsSolutionSetup = (): Tree => {
  vi.spyOn(solutionSetup, 'isUsingTsSolutionSetup').mockImplementation(
    () => true,
  );

  const tree = createTreeWithEmptyWorkspace();

  tree.write('pnpm-workspace.yaml', `packages:\n  - packages/*`);

  tree.write('tsconfig.json', '{}');
  return tree;
};

/**
 * Snapshot all files within a directory in the given tree
 */
export const snapshotTreeDir = (tree: Tree, dir: string) => {
  if (tree.isFile(dir)) {
    expect(tree.read(dir, 'utf-8')).toMatchSnapshot(dir);
  } else {
    tree
      .children(dir)
      .forEach((subDir) =>
        snapshotTreeDir(tree, joinPathFragments(dir, subDir)),
      );
  }
};
