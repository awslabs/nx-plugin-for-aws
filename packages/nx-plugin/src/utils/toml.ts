/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import TOML from '@iarna/toml';

/**
 * Update a toml file
 */
export const updateToml = (
  tree: Tree,
  filePath: string,
  updater: (prev: TOML.JsonMap) => TOML.JsonMap,
) => {
  const prev = readToml(tree, filePath);
  tree.write(filePath, TOML.stringify(updater(prev)));
};

/**
 * Read a toml file
 */
export const readToml = (tree: Tree, filePath: string): TOML.JsonMap => {
  if (!tree.exists(filePath)) {
    throw new Error(`${filePath} does not exist`);
  }
  return TOML.parse(tree.read(filePath, 'utf-8'));
};
