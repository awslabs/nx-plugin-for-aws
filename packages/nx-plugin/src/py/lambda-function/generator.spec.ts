/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import {
  LAMBDA_FUNCTION_GENERATOR_INFO,
  lambdaFunctionProjectGenerator,
} from './generator';
import { parse } from '@iarna/toml';
import {
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
} from '../../utils/shared-constructs-constants';
import { joinPathFragments } from '@nx/devkit';
import { sortObjectKeys } from '../../utils/object';
import { expectHasMetricTags } from '../../utils/metrics.spec';
import { UVPyprojectToml } from '@nxlv/python/src/provider/uv/types';

describe('lambda-handler project generator', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
  });

  it('should generate a lambda function in a project with correct structure', async () => {
    // Setup a Python project
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {},
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    expect(
      tree.exists('apps/test_project/test_project/test_function.py'),
    ).toBeTruthy();
    expect(
      tree.exists('apps/test_project/tests/test_test_function.py'),
    ).toBeTruthy();
  });

  it('should set up project configuration with Lambda Function targets', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    const projectConfig = JSON.parse(
      tree.read('apps/test_project/project.json', 'utf-8'),
    );

    // Verify Lambda Function-specific targets
    expect(projectConfig.targets.bundle).toBeDefined();
    expect(projectConfig.targets.bundle.outputs).toEqual([
      '{workspaceRoot}/dist/apps/test_project/bundle',
    ]);
    expect(projectConfig.targets.bundle.options.commands).toContain(
      'uv export --frozen --no-dev --no-editable --project test_project -o dist/apps/test_project/bundle/requirements.txt',
    );

    // Verify build dependencies
    expect(projectConfig.targets.build.dependsOn).toContain('bundle');

    const pyprojectToml = parse(
      tree.read('apps/test_project/pyproject.toml', 'utf-8'),
    ) as UVPyprojectToml;

    // Verify project dependencies
    expect(pyprojectToml.project.dependencies).toContain(
      'aws-lambda-powertools',
    );
    expect(pyprojectToml.project.dependencies).toContain(
      'aws-lambda-powertools[tracer]',
    );
  });

  it('should set up shared constructs for Lambda Handler', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    // Verify shared constructs files
    const lambdaHandlerPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'test-function.ts',
    );

    expect(tree.exists(lambdaHandlerPath)).toBeTruthy();

    // Verify exports in index files
    const lambdaHandlersIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'index.ts',
    );
    expect(tree.read(lambdaHandlersIndexPath, 'utf-8')).toContain(
      './test-function.js',
    );

    const appIndexPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'index.ts',
    );
    expect(tree.read(appIndexPath, 'utf-8')).toContain(
      './lambda-functions/index.js',
    );
  });

  it('should update shared constructs build dependencies', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    const sharedConstructsConfig = JSON.parse(
      tree.read(
        joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR, 'project.json'),
        'utf-8',
      ),
    );

    expect(sharedConstructsConfig.targets.build.dependsOn).toContain(
      'proj.test_function:build',
    );
  });

  it('should handle custom directory path', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      functionPath: 'nested/path',
      eventType: 'Any',
    });

    expect(
      tree.exists(
        'apps/test_project/test_project/nested/path/test_function.py',
      ),
    ).toBeTruthy();
  });

  it('should generate Lambda Function construct with correct class name', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    const lambdaFunctionPath = joinPathFragments(
      PACKAGES_DIR,
      SHARED_CONSTRUCTS_DIR,
      'src',
      'app',
      'lambda-functions',
      'test-function.ts',
    );
    const lambdaFunctionContent = tree.read(lambdaFunctionPath, 'utf-8');

    expect(lambdaFunctionContent).toContain(
      'export class TestFunction extends Function',
    );
  });

  it('should handle custom event type in python handler file', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'APIGatewayProxyEvent',
    });

    const lambdaFunctionContent = tree.read(
      'apps/test_project/test_project/test_function.py',
      'utf-8',
    );

    expect(lambdaFunctionContent).toContain(
      'from aws_lambda_powertools.utilities.data_classes import APIGatewayProxyEvent, event_source',
    );
    expect(lambdaFunctionContent).toContain(
      '@event_source(data_class=APIGatewayProxyEvent)',
    );
    expect(lambdaFunctionContent).toContain(
      'def lambda_handler(event: APIGatewayProxyEvent, context: LambdaContext)',
    );
  });

  it('should match snapshot', async () => {
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    const appChanges = sortObjectKeys(
      tree
        .listChanges()
        .filter((f) => f.path.endsWith('.py'))
        .reduce((acc, curr) => {
          acc[curr.path] = tree.read(curr.path, 'utf-8');
          return acc;
        }, {}),
    );
    // Verify project metadata
    expect(appChanges).toMatchSnapshot('main-snapshot');
  });

  it('should add generator metric to app.ts', async () => {
    // Call the generator function
    tree.write(
      'apps/test_project/project.json',
      JSON.stringify({
        name: 'test-project',
        root: 'apps/test_project',
        sourceRoot: 'apps/test_project/test_project',
        targets: {
          build: {
            dependsOn: ['lint', 'compile', 'test'],
            options: {
              outputPath: '{workspaceRoot}/dist/apps/test_project',
            },
          },
        },
      }),
    );

    tree.write(
      'apps/test_project/pyproject.toml',
      `[project]
          dependencies = []
      `,
    );

    await lambdaFunctionProjectGenerator(tree, {
      project: 'test-project',
      functionName: 'test-function',
      eventType: 'Any',
    });

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, LAMBDA_FUNCTION_GENERATOR_INFO.metric);
  });
});
