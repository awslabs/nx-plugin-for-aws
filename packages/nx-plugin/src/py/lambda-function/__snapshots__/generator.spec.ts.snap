// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lambda-handler project generator > should match snapshot > main-snapshot 1`] = `
{
  "apps/test_project/test_project/test_function.py": "import os

from aws_lambda_powertools import Logger, Metrics, Tracer
from aws_lambda_powertools.metrics import MetricUnit
from aws_lambda_powertools.utilities.typing import LambdaContext

os.environ["POWERTOOLS_METRICS_NAMESPACE"] = "TestFunction"
os.environ["POWERTOOLS_SERVICE_NAME"] = "TestFunction"

logger: Logger = Logger()
metrics: Metrics = Metrics()
tracer: Tracer = Tracer()

@tracer.capture_lambda_handler
@metrics.log_metrics
def lambda_handler(event: dict, context: LambdaContext):
    logger.info("Received event", extra={"event": event })
    metrics.add_metric(name="InvocationCount", unit=MetricUnit.Count, value=1)

    try:
        # TODO: Implement
        metrics.add_metric(name="SuccessCount", unit=MetricUnit.Count, value=1)
        # TODO: Implement success response if required
    except Exception as e:
        logger.exception(e)
        metrics.add_metric(name="ErrorCount", unit=MetricUnit.Count, value=1)
        # TODO: Implement error response if required
",
  "apps/test_project/tests/test_test_function.py": "def test_handler():
    pass
",
  "packages/common/constructs/src/app/lambda-functions/index.ts": "export * from './test-project-test-function.js';
",
  "packages/common/constructs/src/app/lambda-functions/test-project-test-function.ts": "import { Construct } from 'constructs';
import * as url from 'url';
import { Code, Function, Runtime, Tracing } from 'aws-cdk-lib/aws-lambda';
import { Duration } from 'aws-cdk-lib';

export class TestProjectTestFunction extends Function {
  constructor(scope: Construct, id: string) {
    super(scope, id, {
      timeout: Duration.seconds(30),
      runtime: Runtime.PYTHON_3_12,
      handler: 'test_project.test_function.lambda_handler',
      code: Code.fromAsset(
        url.fileURLToPath(
          new URL(
            '../../../../../../dist/apps/test_project/bundle',
            import.meta.url,
          ),
        ),
      ),
      tracing: Tracing.ACTIVE,
      environment: {
        AWS_CONNECTION_REUSE_ENABLED: '1',
      },
    });
  }
}
",
}
`;
